### PR TITLE
add decorrelated batchnormalization (take-over)

### PR DESCRIPTION
**Pull request by crcrpar**
take over #4686 

Both `F.decorrelated_batch_normalization` and `F.fixed_decorrelated_batch_normalization` do no support double backward. (rel: #4449)

### CODE REVIEW

**Comment by crcrpar** [comment]
I changed DecorreleatedBatchNormalization link to support `config.in_recomputing`. ref:https://github.com/chainer/chainer/pull/5557

**Comment by crcrpar** [comment]
@hvy could you give me another review?

**Comment by hvy** [comment]
Sorry for the late review. I'll try taking a look as soon as possible.

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        if g > 1:
+            gx = gx.reshape((-1, c, ) + gy.shape[2:])
+
+        self.retain_outputs(())
+        return gx,
+
+    def backward(self, inputs, grad_outputs):
+        # TODO(crcrpar): Implement this.
+        raise NotImplementedError("Double backward is not implemented for"
+                                  " decorrelated batch normalizatin.")
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Can you change to single quotes as per convention? There's also a typo `normalizatin`.

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        rhs = x_hat - mean[Ellipsis, None]
+        gprojection = (x_hat - rhs).T.dot(gy_hat)
+        gmean = -gx[:, 0]
+        self.retain_outputs(())
+        return gx, gmean, gprojection
+
+    def backward(self, inputs, grad_outputs):
+        # TODO(crcrpar): Implement this.
+        raise NotImplementedError("Double backward is not implemented for"
+                                  " fixed decorrelated batch normalizatin.")
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Ditto regarding single quote and typo.

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+    Args:
+        x (:class:`~chainer.Variable`): Input variable.
+        mean (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Shifting parameter of input.
+        projection (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Projection matrix for decorrelation of input.
+        groups (int): Number of groups to use for group whitening.
+
+
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Seems like an extra blank line.
```suggestion
```

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+            the decay parameter. If ``None``, the expected mean is initialized
+            to zero.
+        running_projection (:ref:`ndarray`):
+            Expected value of the project matrix. This is a
+            running average of the projection over several mini-batches using
+            the decay parameter. If ``None``, the expected projected is
+            initialized to the identity matrix.
+        decay (float): Decay rate of moving average. It is used during
+            training.
+
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Can you write a `Returns:` ?

File: chainer/functions/normalization/decorrelated_batch_normalization.py (hunk 3 above)
Comment (hvy) [reply chainer/functions/normalization/decorrelated_batch_normalization.py]: Ditto regarding returns.

**Comment by crcrpar** [comment]
I did some DRY and added Returns.

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        if g > 1:
+            gx = gx.reshape((-1, c, ) + gy.shape[2:])
+
+        self.retain_outputs(())
+        return gx,
+
+    def backward(self, inputs, grad_outputs):
+        # TODO(crcrpar): Implement this.
+        raise NotImplementedError('Double backward is not implemented for'
+                                  ' decorrelated batch normalizatin.')
```
Comment (toslunar) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]:
```suggestion
                                  ' decorrelated batch normalization.')
```

Comment (crcrpar) [reply chainer/functions/normalization/decorrelated_batch_normalization.py]: Thank you!

**Comment by crcrpar** [comment]
@hvy PTAL

**Comment by hvy** [comment]
I am well aware of this PR but am sorry that I have not been able to take a look at your changes. Are they of urge, if so I might reassign it to someone else? I should be able to take a look at it next week otherwise.

**Comment by crcrpar** [comment]
There's no need to hurry, thanks.


2019年3月26日(火) 2:04 Hiroyuki Vincent Yamazaki <notifications@github.com>:

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+        if g > 1:
+            x = x.reshape((b * g, C) + x.shape[2:])
+        x_hat = x.transpose((1, 0) + spatial_axis).reshape(C, -1)
+
+        mean = x_hat.mean(axis=1)
+        x_hat = x_hat - mean[:, None]
+        self.eps = x.dtype.type(self.eps)
+
+        identity = self.eps * xp.eye(C, dtype=x.dtype)
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: A nitpick but I find this variable name slightly misleading. I assume it's there for numerical stability, how about simply `eps` ?

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+        self.y_hat_pca = U.dot(x_hat)  # PCA whitening
+        y_hat = self.eigvectors.dot(self.y_hat_pca)  # ZCA whitening
+
+        y = y_hat.reshape((C, b * g,) + x.shape[2:]).transpose(
+            (1, 0) + spatial_axis)
+        if self.groups > 1:
+            y = y.reshape((-1, c) + x.shape[2:])
+
+        # Update running statistics
+        q = x.size // C
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Is this `q` not equal to the previously computed `m` ?

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        return gx, gmean, gprojection
+
+    def backward(self, inputs, grad_outputs):
+        # TODO(crcrpar): Implement this.
+        raise NotImplementedError('Double backward is not implemented for'
+                                  ' fixed decorrelated batch normalizatin.')
+
+
+def decorrelated_batch_normalization(x, **kwargs):
+    """Decorrelated batch normalization function.
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Could you document the default values, similar to e.g. https://github.com/chainer/chainer/blob/v5.3.0/chainer/functions/normalization/batch_normalization.py#L597 ?

Comment (hvy) [reply chainer/functions/normalization/decorrelated_batch_normalization.py]: Or actually, do you need `argument.parse_kwarg` at all?

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+        ~chainer.Variable: The output variable which has the same shape as
+        :math:`x`.
+
+    .. seealso::
+       :func:`functions.decorrelated_batch_normalization`,
+       :class:`links.DecorrelatedBatchNormalization`
+
+    """
+    groups, = argument.parse_kwargs(
+        kwargs, ('groups', 16))
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Similar to `batch_normalization` but do you need `argument.parse_kwarg` ?

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+def _calc_axis_and_m(x_shape, batch_size, groups):
+    m = batch_size * groups
+    spatial_ndim = len(x_shape) - 2
+    spatial_axis = tuple(range(2, 2 + spatial_ndim))
+    for i in spatial_axis:
+        m *= x_shape[i]
+    return spatial_axis, m
+
+
+class DecorrelatedBatchNormalizationFunction(function_node.FunctionNode):
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: A detail and a matter of consistency but do you need the "Function" postfix to the class name? Ditto for the fixed one.

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+        self.retain_inputs(())
+        x = inputs[0]
+        xp = backend.get_array_module(x)
+        x_shape = x.shape
+        b, c = x_shape[:2]
+        g = self.groups
+        C = c // g
+        spatial_axis, m = _calc_axis_and_m(x_shape, b, g)
+
+        if self.running_mean is None:
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: If they're not given by the caller, we don't need to allocate/update any running values at all.

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+        rhs = x_hat - mean[Ellipsis, None]
+        gprojection = (x_hat - rhs).T.dot(gy_hat)
+        gmean = -gx[:, 0]
+        self.retain_outputs(())
+        return gx, gmean, gprojection
+
+    def backward(self, inputs, grad_outputs):
+        # TODO(crcrpar): Implement this.
+        raise NotImplementedError('Double backward is not implemented for'
+                                  ' fixed decorrelated batch normalizatin.')
```
Comment (hvy) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: Ops, a minor typo.

File: chainer/links/normalization/decorrelated_batch_normalization.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        dtype (numpy.dtype): Type to use in computing.
+
+    See: `Decorrelated Batch Normalization <https://arxiv.org/abs/1804.08450>`_
+
+    .. seealso::
+       :func:`~chainer.functions.decorrelated_batch_normalization`,
+       :func:`~chainer.functions.fixed_decorrelated_batch_normalization`
+
+    Attributes:
+        avg_mean (numpy.ndarray or cupy.ndarray): Population mean.
```
Comment (hvy) [inline chainer/links/normalization/decorrelated_batch_normalization.py]: Should we use this new style ":ref:`ndarray`"?

File: chainer/links/normalization/decorrelated_batch_normalization.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        avg_projection (numpy.ndarray or cupy.ndarray): Population
```
Comment (hvy) [inline chainer/links/normalization/decorrelated_batch_normalization.py]: Ditto.

File: chainer/links/normalization/decorrelated_batch_normalization.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+        self.avg_projection = numpy.eye(size // groups, dtype=dtype)
+        self.register_persistent('avg_projection')
+        self.N = 0
+        self.register_persistent('N')
+        self.decay = decay
+        self.eps = eps
+        self.groups = groups
+
+    def forward(self, x, **kwargs):
+        """Invokes the forward propagation of DecorrelatedBatchNormalization.
```
Comment (hvy) [inline chainer/links/normalization/decorrelated_batch_normalization.py]: Could you add a signature documentation at the top with default values?

Comment (hvy) [reply chainer/links/normalization/decorrelated_batch_normalization.py]: Maybe we don't need `argument.parse_kwargs` neither?

File: chainer/links/normalization/decorrelated_batch_normalization.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+                avg_projection = None
+
+            ret = functions.decorrelated_batch_normalization(
+                x, groups=self.groups, eps=self.eps,
+                running_mean=avg_mean, running_projection=avg_projection,
+                decay=decay)
+        else:
+            # Use running average statistics or fine-tuned statistics.
+            # mean = variable.Variable(self.avg_mean)
+            # projection = variable.Variable(self.avg_projection)
```
Comment (hvy) [inline chainer/links/normalization/decorrelated_batch_normalization.py]: Are those two lines above carelesses (that should be removed)?

File: tests/chainer_tests/functions_tests/normalization_tests/test_decorrelated_batch_normalization.py [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+    'dtype': [numpy.float32, numpy.float64],
+    'contiguous': ['C', None],
+})))
+@testing.backend.inject_backend_tests(
+    None,
+    # CPU tests
+    [{}]
+    # GPU tests
+    + [{'use_cuda': True}]
+)
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/normalization_tests/test_decorrelated_batch_normalization.py]: It is possible to test ChainerX as well?

File: tests/chainer_tests/functions_tests/normalization_tests/test_decorrelated_batch_normalization.py [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+    'dtype': [numpy.float16, numpy.float32, numpy.float64],
+    'contiguous': ['C', None],
+})))
+@testing.backend.inject_backend_tests(
+    None,
+    # CPU tests
+    [{}]
+    # GPU tests
+    + [{'use_cuda': True}]
+)
```
Comment (hvy) [inline tests/chainer_tests/functions_tests/normalization_tests/test_decorrelated_batch_normalization.py]: Ditto regarding ChainerX.

**Comment by hvy** [comment]
@crcrpar sorry for the late comments. I wrote some. Could you take a look?

File: chainer/links/normalization/decorrelated_batch_normalization.py [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+
+    See: `Decorrelated Batch Normalization <https://arxiv.org/abs/1804.08450>`_
+
+    .. seealso::
+       :func:`~chainer.functions.decorrelated_batch_normalization`,
+       :func:`~chainer.functions.fixed_decorrelated_batch_normalization`
+
+    Attributes:
+        avg_mean (:ref:`ndarray`): Population mean.
+        avg_projection (:ref:`ndarray): Population
```
Comment (hvy) [inline chainer/links/normalization/decorrelated_batch_normalization.py]: Ops, missing a closing backtick.
```suggestion
        avg_projection (:ref:`ndarray`): Population
```

Comment (crcrpar) [reply chainer/links/normalization/decorrelated_batch_normalization.py]: Thank you

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 20]
```diff
@@ -0,0 +1,303 @@
...
+        return gx, gmean, gprojection
+
+    def backward(self, inputs, grad_outputs):
+        # TODO(crcrpar): Implement this.
+        raise NotImplementedError('Double backward is not implemented for'
+                                  ' fixed decorrelated batch normalization.')
+
+
+def decorrelated_batch_normalization(x, **kwargs):
+    """decorrelated_batch_normalization(x, groups=16, eps2e-5,
```
Comment (toslunar) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]:
```suggestion
    """decorrelated_batch_normalization(x, *, groups=16, eps2e-5,
```

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 21]
```diff
@@ -0,0 +1,303 @@
...
+            kwargs, ('groups', 16), ('eps', 2e-5), ('running_mean', None),
+            ('running_projection', None), ('decay', 0.9))
+
+    f = DecorrelatedBatchNormalization(
+        groups, eps, running_mean, running_projection, decay)
+    return f.apply((x,))[0]
+
+
+def fixed_decorrelated_batch_normalization(x, mean, projection, groups=16):
+    """fixed_decorrelated_batch_normalization(x, mean, projection, groups=16)
```
Comment (toslunar) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]: The line is unnecessary because the signature is the same.

File: chainer/links/normalization/decorrelated_batch_normalization.py [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+        self.avg_projection = numpy.eye(size // groups, dtype=dtype)
+        self.register_persistent('avg_projection')
+        self.N = 0
+        self.register_persistent('N')
+        self.decay = decay
+        self.eps = eps
+        self.groups = groups
+
+    def forward(self, x, **kwargs):
+        """forward(self, x, finetune=False)
```
Comment (toslunar) [inline chainer/links/normalization/decorrelated_batch_normalization.py]:
```suggestion
        """forward(self, x, *, finetune=False)
```

File: chainer/functions/normalization/decorrelated_batch_normalization.py (hunk 21 above)
Comment (crcrpar) [reply chainer/functions/normalization/decorrelated_batch_normalization.py]: Thank you!!!

File: chainer/functions/normalization/decorrelated_batch_normalization.py [hunk 23]
```diff
@@ -0,0 +1,301 @@
...
+        return gx, gmean, gprojection
+
+    def backward(self, inputs, grad_outputs):
+        # TODO(crcrpar): Implement this.
+        raise NotImplementedError('Double backward is not implemented for'
+                                  ' fixed decorrelated batch normalization.')
+
+
+def decorrelated_batch_normalization(x, **kwargs):
+    """decorrelated_batch_normalization(x, *, groups=16, eps2e-5,
```
Comment (toslunar) [inline chainer/functions/normalization/decorrelated_batch_normalization.py]:
```suggestion
    """decorrelated_batch_normalization(x, *, groups=16, eps2e-5, \
```

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Travis failure seems unrelated. Thanks for the long running PR, let me merge this one! LGTM!
